examples/general/basic_dead_store/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+#![feature(rustc_private)]
+#![warn(unused_extern_crates)]
+
+extern crate rustc_ast;
+extern crate rustc_hir;
+extern crate rustc_span;
+use clippy_utils::{diagnostics::span_lint_and_help, get_parent_expr};
+
+use rustc_ast::LitKind;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_span::Span;
+
+dylint_linting::impl_late_lint! {
+    /// ### What it does
+    /// Finds instances of dead stores in arrays: array positions that are assigned twice without a
+    ///  use or read in between.
+    ///
+    /// ### Why is this bad?
+    /// A dead store might indicate a logic error in the program or an unnecessary assignment.
+    ///
+    /// ### Known problems
+    /// This lint only checks for literal indices and will not try to find instances where an array
+    /// is indexed by a variable.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let mut arr = [0u64; 2];
+    /// arr[0] = 1;
+    /// arr[0] = 2;
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let mut arr = [0u64; 2];
+    /// arr[0] = 2;
+    /// arr[1] = 1;
+    /// ```
+    pub BASIC_DEAD_STORE,
+    Warn,
+    "An array element is assigned twice without a use or read in between",
+    BasicDeadStore::default()
+}
+
+#[derive(Default)]
+pub struct BasicDeadStore {
+    /// Stores instances of array-indexing with literal (array name, index, span)
+    arr_and_idx_vec: Vec<(String, u128, Span)>,
+}
+
+impl BasicDeadStore {
+    /// Removes all stored values of the given array
+    fn clear_stores_of(&mut self, string: &String) {
+        self.arr_and_idx_vec
+            .retain(|(arr_string, _idx, _span)| arr_string != string);
+    }
+
+    /// Returns all instances where the given array is indexed at `idx`
+    fn get_pairs_with_same_name_idx(
+        &self,
+        string: &String,
+        idx: &u128,
+    ) -> Vec<&(String, u128, Span)> {
+        self.arr_and_idx_vec
+            .iter()
+            .filter(|(arr_string, arr_idx, _span)| arr_string == string && arr_idx == idx)
+            .collect()
+    }
+}
+
+/// Checks if the given expression is an assignment to an array indexed by a literal.
+/// Returns the tuple (array name, indexed position, span)
+fn is_assignment_to_array_indexed_by_literal(
+    expr: &Expr,
+    arr_string: &String,
+    tcx: &LateContext<'_>,
+) -> Option<(String, u128, Span)> {
+    let index_expr = get_parent_expr(tcx, expr)?;
```

Comment:
If I'm understanding correctly, this function matches an index expression, then it matches _upward_ to see if it's in an assignment, then it matches _downward_ to see if the index is a literal. Is that right?

Would it make more sense to match the assignment and then just match downward?

I'm not sure if this is helpful: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=98acb5ca4f29b9d2be7d69fece5a77ed

Reply:
Yes, currently, we start with expr and need to check all other 3 nodes:
```
       assign_expr is ExprKind::Assign
         /
     index_expr is ExprKind::Index
   /         \
expr        lit is ExprKind::Lit
```
Matching on `expr` and going up and down in the tree seemed better:
Let's suppose I found all assignments to arrays with literals. Then, we would need to determine if there are any uses of these arrays between the assignments to the same position.
To determine if a Path expression of the same array was a `use`, we would need to check that it wasn't the original assignment and that it was in the proper location (i.e., between the previously collected assignments). This seemed much more complex than navigating at the path level.

Navigating at the Path level meant that I would never see the same array reference twice, and the check_expr return order seemed to match the program order.

Reply:
> To determine if a Path expression of the same array was a `use`, we would need to check that it wasn't the original assignment

OK. I see what you are saying.